File: lecture_summarizer/ingest/drive.py
from __future__ import annotations

import os
import tempfile
from pathlib import Path
from typing import Protocol

from ..core.types import Asset, Job, SourceKind


class DriveDownloader(Protocol):
    def download(self, file_id: str, destination: Path) -> Path: ...


class DriveIngestor:
    """Download files referenced by drive:// identifiers."""

    def __init__(self, *, cache_dir: Path | None = None, downloader: DriveDownloader | None = None) -> None:
        self._cache_dir = Path(cache_dir or Path(tempfile.gettempdir()) / "lecture-drive-cache")
        self._cache_dir.mkdir(parents=True, exist_ok=True)
        self._downloader = downloader

    def discover(self, job: Job) -> list[Asset]:
        source = job.source
        if source.startswith("drive://"):
            file_id = source.split("drive://", 1)[1]
        elif source.startswith("gdrive://"):
            file_id = source.split("gdrive://", 1)[1]
        else:
            return []
        file_id = file_id.strip()
        if not file_id:
            raise ValueError("Drive URI missing file identifier")

        destination = self._cache_dir / file_id
        if destination.exists():
            return [self._to_asset(destination)]

        downloader = self._resolve_downloader()
        downloaded = downloader.download(file_id, destination)
        return [self._to_asset(downloaded)]

    def _resolve_downloader(self) -> DriveDownloader:
        if self._downloader is not None:
            return self._downloader
        try:
            from googleapiclient.discovery import build  # type: ignore
            from google.oauth2 import service_account  # type: ignore
        except ImportError as exc:  # pragma: no cover - optional dependency
            raise RuntimeError("google-api-python-client is required for Drive ingestion") from exc

        credentials_path = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")
        if not credentials_path:
            raise RuntimeError("Set GOOGLE_APPLICATION_CREDENTIALS to use Drive ingestion")

        scopes = ["https://www.googleapis.com/auth/drive.readonly"]
        creds = service_account.Credentials.from_service_account_file(credentials_path, scopes=scopes)
        service = build("drive", "v3", credentials=creds)

        class _DefaultDownloader:
            def __init__(self, svc, cache: Path):
                self._service = svc
                self._cache = cache

            def download(self, file_id: str, destination: Path) -> Path:
                request = self._service.files().get_media(fileId=file_id)
                from googleapiclient.http import MediaIoBaseDownload  # type: ignore
                import io

                fh = io.BytesIO()
                downloader = MediaIoBaseDownload(fh, request)
                done = False
                while not done:
                    _, done = downloader.next_chunk()
                fh.seek(0)
                destination.parent.mkdir(parents=True, exist_ok=True)
                with destination.open("wb") as fp:
                    fp.write(fh.read())
                return destination

        return _DefaultDownloader(service, self._cache_dir)

    @staticmethod
    def _to_asset(path: Path) -> Asset:
        mime = "application/octet-stream"
        if path.suffix.lower() == ".pdf":
            media = "pdf"
            mime = "application/pdf"
        elif path.suffix.lower() in {".png", ".jpg", ".jpeg", ".gif"}:
            media = "image"
            mime = f"image/{path.suffix.lower().lstrip('.')}"
        elif path.suffix.lower() in {".mp4", ".mov"}:
            media = "video"
            mime = "video/mp4"
        elif path.suffix.lower() in {".mp3", ".wav", ".m4a"}:
            media = "audio"
            mime = "audio/mpeg"
        else:
            media = "pdf"
        return Asset(path=path, media=media, source_kind=SourceKind.DRIVE, mime=mime)

File: src/ingest/drive.rs
use std::fs::{self, File};
use std::io::copy;
use std::path::{Path, PathBuf};

use anyhow::{anyhow, bail, Context, Result};
use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
use reqwest::blocking::Client;
use serde::{Deserialize, Serialize};
use time::{Duration, OffsetDateTime};

use crate::core::{Asset, Job, SourceKind};
use crate::utils::ensure_dir;
use crate::video::sha256sum;

const SCOPE: &str = "https://www.googleapis.com/auth/drive.readonly";

#[derive(Debug, Clone)]
pub struct DriveIngestor {
    cache_dir: PathBuf,
    client: Client,
}

impl DriveIngestor {
    pub fn new(cache_dir: Option<PathBuf>) -> Result<Self> {
        let cache = cache_dir.unwrap_or_else(|| std::env::temp_dir().join("recapit-drive-cache"));
        ensure_dir(&cache)?;
        Ok(Self {
            cache_dir: cache,
            client: Client::builder()
                .timeout(std::time::Duration::from_secs(120))
                .build()?,
        })
    }

    pub fn discover(&self, job: &Job) -> Result<Vec<Asset>> {
        let source = job.source.trim();
        let file_id = if let Some(rest) = source.strip_prefix("drive://") {
            rest
        } else if let Some(rest) = source.strip_prefix("gdrive://") {
            rest
        } else {
            return Ok(vec![]);
        }
        .trim();
        if file_id.is_empty() {
            bail!("Drive URI missing file identifier");
        }
        let destination = self.cache_dir.join(file_id);
        if !destination.exists() {
            let creds = ServiceAccountCredentials::load_from_env()?;
            let token = creds.fetch_token(&self.client)?;
            self.download_file(file_id, &destination, &token)?;
        }

        let media = infer_media(&destination);
        let mime = guess_mime(&destination);
        let meta = serde_json::json!({
            "drive_file_id": file_id,
            "sha256": sha256sum(&destination)?,
            "size_bytes": destination.metadata().ok().map(|m| m.len()),
        });
        Ok(vec![Asset {
            path: destination,
            media: media.into(),
            page_index: None,
            source_kind: SourceKind::Drive,
            mime: Some(mime.into()),
            meta,
        }])
    }

    fn download_file(&self, file_id: &str, destination: &Path, token: &str) -> Result<()> {
        ensure_dir(destination.parent().unwrap_or_else(|| Path::new(".")))?;
        let url = format!("https://www.googleapis.com/drive/v3/files/{file_id}?alt=media");
        let mut response = self
            .client
            .get(url)
            .bearer_auth(token)
            .send()
            .with_context(|| format!("Downloading Drive file {file_id}"))?;
        if !response.status().is_success() {
            bail!("Drive download failed with status {}", response.status());
        }
        let temp = destination.with_extension("part");
        let mut file = File::create(&temp)?;
        copy(&mut response, &mut file)?;
        fs::rename(temp, destination)?;
        Ok(())
    }
}

#[derive(Debug, Deserialize)]
struct ServiceAccountCredentials {
    client_email: String,
    private_key: String,
}

impl ServiceAccountCredentials {
    fn load_from_env() -> Result<Self> {
        let path = std::env::var("GOOGLE_APPLICATION_CREDENTIALS")
            .map_err(|_| anyhow!("Set GOOGLE_APPLICATION_CREDENTIALS to use Drive ingestion"))?;
        let text = std::fs::read_to_string(&path)
            .with_context(|| format!("Reading service account credentials {path}"))?;
        let creds: ServiceAccountCredentials = serde_json::from_str(&text)?;
        Ok(creds)
    }

    fn fetch_token(&self, client: &Client) -> Result<String> {
        let now = OffsetDateTime::now_utc();
        let claim = ServiceAccountClaim {
            iss: &self.client_email,
            scope: SCOPE,
            aud: "https://oauth2.googleapis.com/token",
            exp: now.saturating_add(Duration::minutes(55)).unix_timestamp(),
            iat: now.unix_timestamp(),
        };
        let jwt = encode(
            &Header::new(Algorithm::RS256),
            &claim,
            &EncodingKey::from_rsa_pem(self.private_key.as_bytes())?,
        )?;
        let params = [
            ("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"),
            ("assertion", jwt.as_str()),
        ];
        let resp = client
            .post("https://oauth2.googleapis.com/token")
            .form(&params)
            .send()?;
        if !resp.status().is_success() {
            bail!("OAuth token exchange failed with status {}", resp.status());
        }
        let payload: TokenResponse = resp.json()?;
        Ok(payload.access_token)
    }
}

#[derive(Serialize)]
struct ServiceAccountClaim<'a> {
    iss: &'a str,
    scope: &'a str,
    aud: &'a str,
    exp: i64,
    iat: i64,
}

#[derive(Debug, Deserialize)]
struct TokenResponse {
    access_token: String,
}

fn infer_media(path: &Path) -> &'static str {
    let ext = path
        .extension()
        .map(|ext| ext.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    match ext.as_str() {
        "pdf" => "pdf",
        "png" | "jpg" | "jpeg" | "gif" => "image",
        "mp4" | "mov" | "mkv" => "video",
        "mp3" | "wav" | "m4a" => "audio",
        _ => "pdf",
    }
}

fn guess_mime(path: &Path) -> &'static str {
    let ext = path
        .extension()
        .map(|ext| ext.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    match ext.as_str() {
        "pdf" => "application/pdf",
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "gif" => "image/gif",
        "mp4" | "mov" | "mkv" => "video/mp4",
        "mp3" | "wav" | "m4a" => "audio/mpeg",
        _ => "application/octet-stream",
    }
}
